lib/swoosh/adapters/customer_io.ex

```diff
@@ -0,0 +1,214 @@
+      {:ok, %{id: data["delivery_id"]}}
+    else
+      {:ok, code, _headers, body} when code >= 400 ->
+        case Swoosh.json_library().decode(body) do
+          {:ok, error} -> {:error, {code, error}}
+          {:error, _} -> {:error, {code, body}}
+        end
+
+      {:error, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp base_url(config), do: config[:base_url] || @base_url
+
+  defp prepare_body(email) do
+    %{}
+    |> prepare_from(email)
+    |> prepare_to(email)
+    |> prepare_bcc(email)
+    |> prepare_subject(email)
+    |> prepare_content(email)
+    |> prepare_attachments(email)
+    |> prepare_reply_to(email)
+    |> prepare_custom_headers(email)
+    |> prepare_provider_options_body_fields(email)
+  end
+
+  defp prepare_from(body, %Email{from: nil}), do: body
+  defp prepare_from(body, %Email{from: from}), do: Map.put(body, :from, email_item(from))
+
+  defp prepare_to(body, %Email{to: to}),
+    do: Map.put(body, :to, Enum.map_join(to, ",", &email_item(&1)))
+
+  defp prepare_bcc(body, %Email{bcc: []}), do: body
+
+  defp prepare_bcc(body, %Email{bcc: bcc}),
+    do: Map.put(body, :bcc, Enum.map_join(bcc, ",", &email_item(&1)))
+
+  defp prepare_reply_to(body, %Email{reply_to: nil}), do: body
+
+  defp prepare_reply_to(body, %Email{reply_to: reply_to}),
+    do: Map.put(body, :reply_to, email_item(reply_to))
+
+  defp prepare_subject(body, %Email{subject: nil}), do: body
+  defp prepare_subject(body, %Email{subject: subject}), do: Map.put(body, :subject, subject)
+
+  defp prepare_content(body, %Email{html_body: nil, text_body: nil}), do: body
+
+  defp prepare_content(body, %Email{html_body: nil, text_body: text_body}),
+    do: Map.put(body, :plaintext_body, text_body)
+
+  defp prepare_content(body, %Email{html_body: html_body, text_body: nil}),
+    do: Map.put(body, :body, html_body)
+
+  defp prepare_content(body, %Email{html_body: html_body, text_body: text_body}),
+    do: Map.merge(body, %{body: html_body, plaintext_body: text_body})
+
+  defp prepare_attachments(body, %Email{attachments: []}), do: body
+
+  defp prepare_attachments(body, %Email{attachments: attachments}) do
+    attachments =
+      Enum.map(attachments, &%{&1.filename => Swoosh.Attachment.get_content(&1, :base64)})
+
+    Map.put(body, :attachments, attachments)
+  end
+
+  defp prepare_custom_headers(body, %Email{headers: headers})
+       when map_size(headers) == 0,
+       do: body
+
+  defp prepare_custom_headers(body, %Email{headers: headers}) do
+    headers = Enum.map(headers, fn {header, value} -> %{header => value} end)
+    Map.put(body, :headers, headers)
+  end
+
+  defp prepare_provider_options_body_fields(body, %Email{provider_options: provider_options}) do
+    Map.merge(body, Map.take(provider_options, @provider_options_body_fields))
+  end
+
+  defp email_item({"", address}), do: address
+  defp email_item({name, address}), do: name <> " <" <> address <> ">"
+  defp email_item(address), do: address
```

Comment:
https://github.com/swoosh/swoosh/blob/main/lib/swoosh/email/render.ex

Does this work?

Reply:
Yes, I just copied the code over from the Sendgrid adapter, fixed.